examples_tests/20.Megatexture/main.cpp

```diff
@@ -76,359 +51,87 @@ layout (push_constant) uniform Block {
-        Ns = textureVT(PC.map_Ns_data, UV, dUV).x;
-    else
-#endif
-        Ns = PC.Ns;
-
-    vec3 Ni = vec3(PC.Ni);
-
-    vec3 diff = irr_glsl_lambertian_cos_eval(params) * Kd * (1.0-irr_glsl_fresnel_dielectric(Ni,params.NdotL)) * (1.0-irr_glsl_fresnel_dielectric(Ni,params.NdotV));
-    diff *= irr_glsl_diffuseFresnelCorrectionFactor(Ni, Ni*Ni);
-    switch (PC.extra&ILLUM_MODEL_MASK)
-    {
-    case 0:
-        color = vec3(0.0);
-        break;
-    case 1:
-        color = diff;
-        break;
-    case 2:
-    case 3://2 + IBL
-    case 5://basically same as 3
-    case 8://basically same as 5
-    {
-        vec3 spec = Ks*irr_glsl_blinn_phong_fresnel_dielectric_cos_eval(params, Ns, Ni);
-        color = (diff + spec);
-    }
-        break;
-    case 4:
-    case 6:
-    case 7:
-    case 9://basically same as 4
-    {
-        vec3 spec = Ks*irr_glsl_blinn_phong_fresnel_dielectric_cos_eval(params, Ns, Ni);
-        color = spec;
-    }
-        break;
-    default:
-        break;
-    }
 
-    return color;
-}
-#define _IRR_BSDF_COS_EVAL_DEFINED_
+//irr/builtin/glsl/virtual_texturing/functions.glsl/...
+#include <%s>
 )";
-constexpr const char* GLSL_COMPUTE_LIGHTING_OVERRIDE =
-R"(
-vec3 irr_computeLighting(out irr_glsl_ViewSurfaceInteraction out_interaction, in mat2 dUV)
-{
-    irr_glsl_ViewSurfaceInteraction interaction = irr_glsl_calcFragmentShaderSurfaceInteraction(vec3(0.0), ViewPos, Normal);
 
-#ifndef _NO_UV
-    if ((PC.extra&map_bump_MASK) == map_bump_MASK)
-    {
-        interaction.N = normalize(interaction.N);
+constexpr const char* GLSL_FS_PROLOGUE = R"(
+dUV = mat2(dFdx(UV),dFdy(UV));//writing to global var\n
+#define _IRR_FRAG_PROLOGUE_DEFINED_
+)";
 
-        float h = textureVT(PC.map_bump_data, UV, dUV).x;
+using STextureData = asset::ICPUVirtualTexture::STextureData;
 
-        vec2 dHdScreen = vec2(dFdx(h), dFdy(h));
-        interaction.N = irr_glsl_perturbNormal_heightMap(interaction.N, interaction.V.dPosdScreen, dHdScreen);
-    }
-#endif
-    irr_glsl_BSDFIsotropicParams params = irr_glsl_calcBSDFIsotropicParams(interaction, -ViewPos);
-
-    vec3 Ka;
-    switch ((PC.extra&ILLUM_MODEL_MASK))
-    {
-    case 0:
-    {
-#ifndef _NO_UV
-    if ((PC.extra&(map_Kd_MASK)) == (map_Kd_MASK))
-        Ka = textureVT(PC.map_Kd_data, UV, dUV).rgb;
-    else
-#endif
-        Ka = PC.Kd;
-    }
-    break;
-    default:
-#define Ia 0.1
-    {
-#ifndef _NO_UV
-    if ((PC.extra&(map_Ka_MASK)) == (map_Ka_MASK))
-        Ka = textureVT(PC.map_Ka_data, UV, dUV).rgb;
-    else
-#endif
-        Ka = PC.Ka;
-    Ka *= Ia;
-    }
-#undef Ia
-    break;
-    }
+constexpr uint32_t PGTAB_SZ_LOG2 = 7u;
+constexpr uint32_t PGTAB_LAYERS_PER_FORMAT = 1u;
+constexpr uint32_t PGTAB_LAYERS = 3u;
+constexpr uint32_t PAGE_SZ_LOG2 = 7u;
+constexpr uint32_t TILES_PER_DIM_LOG2 = 6u;
+constexpr uint32_t PHYS_ADDR_TEX_LAYERS = 3u;
+constexpr uint32_t PAGE_PADDING = 8u;
+constexpr uint32_t MAX_ALLOCATABLE_TEX_SZ_LOG2 = 12u; //4096
```

Comment:
the VirtualTexture class should compute this for you from the `PGTAB_SZ_LOG2`
OR
is should compute the `PGTAB_SZ_LOG2` given `MAX_ALLOCATABLE_TEX_SZ_LOG2 `

Reply:
the only reason it exist is to not create pagetable mip levels that will never be used, so the second option makes most sense

Reply:
ok, then compute the PGTAB_SZ_LOG2 given MAX_ALLOCATABLE_TEX_SZ_LOG2